pattern/reduction.cpp

```diff
@@ -48,30 +48,43 @@ class Reduction
   void submit_ndrange(std::vector<cl::sycl::event>& events){
     this->submit([this, &events](sycl::buffer<T, 1> *input, sycl::buffer<T, 1> *output,
                         const size_t reduction_size, const size_t num_groups) {
-      events.push_back(this->local_reduce_ndrange(input, output, reduction_size, num_groups));
+      auto ev = this->local_reduce_ndrange(input, output, reduction_size,
+                                            num_groups);
+      // We need to wait for the event here to make sure the kernel finished
+      // execution before swapping the buffers.
+      ev.wait();
+      events.push_back(ev);
     });
   }
 
   void submit_hierarchical(std::vector<cl::sycl::event>& events){
     this->submit([this, &events](sycl::buffer<T, 1> *input, sycl::buffer<T, 1> *output,
                         const size_t reduction_size, const size_t num_groups) {
-      events.push_back(this->local_reduce_hierarchical(input, output, reduction_size,
-                                      num_groups));
+      auto ev = this->local_reduce_hierarchical(input, output, reduction_size,
+                                                  num_groups);
+      // We need to wait for the event here to make sure the kernel finished
+      // execution before swapping the buffers.
+      ev.wait();
+      events.push_back(ev);
     });
   }
 
   bool verify(VerificationSetting &ver) {
-    T result = _final_output_buff->template get_access<sycl::access::mode::read>(
-        sycl::range<1>{0}, sycl::id<1>{1})[0];
+    T result = _final_output_buff->get_host_access()[0];
 
     // Calculate CPU result in fp64 to avoid obtaining a wrong verification result
+    std::vector<T> initial_input(_input.size());
     std::vector<double> input_fp64(_input.size());
-    for(std::size_t i = 0; i < _input.size(); ++i)
-      input_fp64[i] = static_cast<double>(_input[i]);
+    // _input_buff will be re-used as output buffer in the multi-step reduction,
+    // overriding the content of _input. Initialize again with the original
```

Comment:
Why? That should not happen I think. `_input` is only used to initialize the buffer. The `buffer` is of type `T` and initialized with a `const T*`. I don't think a writeback to const pointer from a non-const buffer can happen, as that would break constness. (and if so, we should maybe just call `set_writeback(false)` ).

Reply:
Thank you. Thinking more about it, probably it is not an issue with the writeback itself. The buffer is a member of the class, and therefore should not have been destroyed at the point of verification. Consequently, writeback should not be involved here. Could it be that DPC++ directly operates on input pointers if the device is a CPU, regardless of whether the input point is const or non-const? (which would also be illegal)

Reply:
Yes, my suspicion is that DPC++ re-uses the pointer to avoid an additional allocation on the host, but ignored the `const`ness of the pointer. I tried setting `set_write_back(false)` and it didn't help either, so it indeed seems to be not directly related to write-back but rather re-use of the pointer.